mautic/mautic
# Inject a caching solution into the DeviceDetector

#### Description:

While load testing a Mautic installation's page tracking capabilities (POST to /mtc/event), when the server came under a heavy load we found that determining what device was being used took up more than half the request time (2+ seconds).

DeviceDetector::parseDevice was found to be the culprit, specifically the Spyc::loadWith methods. We saw that this part of the code was cache-ready, but no cache was ever injected so it fell back on a StaticCache.

AbstractDeviceParser::getRegexes() specifically is cache ready:

```php
    /**
     * Returns the result of the parsed yml file defined in $fixtureFile
     *
     * @return array
     */
    protected function getRegexes(): array
    {
        if (empty($this->regexList)) {
            $cacheKey        = 'DeviceDetector-' . DeviceDetector::VERSION . 'regexes-' . $this->getName();
            $cacheKey        = (string) \preg_replace('/([^a-z0-9_-]+)/i', '', $cacheKey);
            $this->regexList = $this->getCache()->fetch($cacheKey);

            if (empty($this->regexList)) {
                $this->regexList = $this->getYamlParser()->parseFile(
                    $this->getRegexesDirectory() . DIRECTORY_SEPARATOR . $this->fixtureFile
                );
                $this->getCache()->save($cacheKey, $this->regexList);
            }
        }

        return $this->regexList;
    }
```

By injecting a cache in the DeviceDetectorFactory, this bit of code now caches its getRegexes() across requests and we saw this bottleneck which took 2+ seconds disappear under a heavy load scenario.

#### Steps to test this PR:

Using your stress testing tool of choice, send too many POST calls to /mtc/event.
Using NewRelic you can inspect the trace of the calls that are timing out and see that the getRegexes() code is bottlenecking.

![image](https://user-images.githubusercontent.com/7923739/144648571-69b56ec3-a632-442e-b58d-9720ea878d90.png)

Apply this patch, then test again and the bottleneck disappears.



<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10665"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
